File: Data Structures & Algorithms/serialize-and-deserialize-binary-tree/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Codec {
public:

    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        std::string result{};
        serialize(root, result);
        result.pop_back();
        return result;
    }

    void serialize(TreeNode* root, std::string& result) {
        if (root == nullptr) {
            result.append("#,");
            return;
        }

        result.append(std::to_string(root->val));
        result.append(",");

        serialize(root->left, result);
        serialize(root->right, result);
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        if (data == "") return nullptr;
        int start = 0;
        return deserialize(data, start);
    }

    TreeNode* deserialize(std::string& data, int& start) {
        int end = data.find(',', start);
        if (end == std::string::npos) {
            end = data.size();
        }

        std::string substr = data.substr(start, end - start);
        start = end + 1;
        if (substr == "#") {
            return nullptr;
        }

        int value = std::stoi(substr);
        TreeNode* root = new TreeNode(value);
        root->left = deserialize(data, start);
        root->right = deserialize(data, start);

        return root;
    }
};

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-1.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {
    public int GoodNodes(TreeNode root) {
        return Traversal(root, int.MinValue);
    }

    public int Traversal(TreeNode node, int max) {
        if (node == null) return 0;
        int count = 0;

        if (node.val >= max) {
            count++;
            max = node.val;
        }

        int leftCount = Traversal(node.left, max);
        int rightCount = Traversal(node.right, max);
        return count + leftCount + rightCount;
    }
}
